File: core/src/main/java/com/github/stephenvinouze/shapeview/ShapeView.java
package com.github.stephenvinouze.shapeview;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.widget.RelativeLayout;

/**
 * Created by stephenvinouze on 05/10/16.
 */
public abstract class ShapeView extends RelativeLayout {

    protected Path shapePath;
    protected final Paint shapePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    protected final Paint shapeBorderPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    protected final RectF shapeCurveRect = new RectF();

    protected int color = Color.TRANSPARENT;
    protected int strokeColor = Color.TRANSPARENT;
    protected int strokeSize = 0;

    public ShapeView(Context context) {
        super(context);
    }

    public ShapeView(Context context, AttributeSet attrs) {
        super(context, attrs);

        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.ShapeView, 0, 0);

        setColor(a.getColor(R.styleable.ShapeView_shapeColor, Color.TRANSPARENT));
        setStrokeColor(a.getColor(R.styleable.ShapeView_shapeStrokeColor, Color.TRANSPARENT));
        setStrokeSize(a.getDimensionPixelSize(R.styleable.ShapeView_shapeStrokeWidth, 0));

        a.recycle();

        initView(context);
    }

    protected void initView(Context context) {
        setBackgroundColor(Color.TRANSPARENT);

        shapePaint.setDither(true);
        shapePaint.setStyle(Paint.Style.FILL);
        shapePaint.setColor(color);

        shapeBorderPaint.setDither(true);
        shapeBorderPaint.setStyle(Paint.Style.STROKE);
        shapeBorderPaint.setColor(strokeColor);
        shapeBorderPaint.setStrokeWidth(strokeSize);
    }

    public int getColor() {
        return color;
    }

    public void setColor(int color) {
        this.color = color;
        shapePaint.setColor(color);
        invalidate();
    }

    public int getStrokeColor() {
        return strokeColor;
    }

    public void setStrokeColor(int strokeColor) {
        this.strokeColor = strokeColor;
        shapeBorderPaint.setColor(strokeColor);
        invalidate();
    }

    public int getStrokeSize() {
        return strokeSize;
    }

    public void setStrokeSize(int strokeSize) {
        this.strokeSize = strokeSize;
        shapeBorderPaint.setStrokeWidth(strokeSize);
        invalidate();
    }

}

File: core/src/main/kotlin/com/github/stephenvinouze/shapeview/ShapeView.kt
package com.github.stephenvinouze.shapeview

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.widget.RelativeLayout

/**
 * Created by stephenvinouze on 05/10/16.
 */
abstract class ShapeView : RelativeLayout {

    protected var shapePath: Path? = null
    protected val shapePaint = Paint(Paint.ANTI_ALIAS_FLAG)
    protected val shapeBorderPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    protected val shapeCurveRect = RectF()

    protected var color: Int = Color.TRANSPARENT
        set(value) {
            field = value
            shapePaint.color = value
            invalidate()
        }
    protected var strokeColor: Int = Color.TRANSPARENT
        set(value) {
            field = value
            shapeBorderPaint.color = value
            invalidate()
        }
    protected var strokeSize: Int = 0
        set(value) {
            field = value
            shapeBorderPaint.strokeWidth = value.toFloat()
            invalidate()
        }

    protected var dashOnSize: Int = 0
        set(value) {
            field = value
            applyDashEffect()
            invalidate()
        }

    protected var dashOffSize: Int = 0
        set(value) {
            field = value
            applyDashEffect()
            invalidate()
        }

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        setBackgroundColor(Color.TRANSPARENT)

        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.ShapeView, 0, 0)

        color = a.getColor(R.styleable.ShapeView_shapeColor, Color.TRANSPARENT)
        strokeColor = a.getColor(R.styleable.ShapeView_shapeStrokeColor, Color.TRANSPARENT)
        strokeSize = a.getDimensionPixelSize(R.styleable.ShapeView_shapeStrokeWidth, 0)
        dashOnSize = a.getDimensionPixelSize(R.styleable.ShapeView_shapeDashOnWidth, 0)
        dashOffSize = a.getDimensionPixelSize(R.styleable.ShapeView_shapeDashOffWidth, 0)

        a.recycle()

        shapePaint.isDither = true
        shapePaint.style = Paint.Style.FILL

        shapeBorderPaint.isDither = true
        shapeBorderPaint.style = Paint.Style.STROKE
    }

    private fun applyDashEffect() {
        shapeBorderPaint.pathEffect = DashPathEffect(floatArrayOf(dashOnSize.toFloat(), dashOffSize.toFloat()), 0f)
    }

}
